internal/controller/flow_aggregation.go

```diff
@@ -0,0 +1,270 @@
+				frontElem := sm.FlowCache.queue.Front()
+				var flowTimestamp time.Time
+				var flowKey FlowKey
+				var err error
+
+				// Check the type of flow and get the flow timestamp and flow key.
+				switch f := frontElem.Value.(type) {
+				case *pb.FalcoFlow:
+					flowTimestamp = f.GetTimestamp().AsTime()
+					flowKey, err = sm.createFlowKey(f)
+				case *pb.CiliumFlow:
+					flowTimestamp = f.GetTime().AsTime()
+					flowKey, err = sm.createFlowKey(f)
+				default:
+					break
+				}
+				if err != nil {
+					return err
+				}
+
+				// If the flow is not expired (timestamp is newer than the cutoff), stop evicting.
+				if flowTimestamp.After(expiredCutoff) {
+					break
+				}
+
+				// Evict the expired flow by removing it from the cache and queue.
+				sm.FlowCache.queue.Remove(frontElem)
+				delete(sm.FlowCache.cache, flowKey)
+
+				sm.sendNetworkFlowRequest(logger, frontElem.Value)
+			}
+
+			// Reset the timer to expire based on the next flow's timestamp.
+			if front := sm.FlowCache.queue.Front(); front != nil {
+				switch f := front.Value.(type) {
+				case *pb.FalcoFlow:
+					resetTimer(timer, f.GetTimestamp().AsTime(), activeTimeout)
+				case *pb.CiliumFlow:
+					resetTimer(timer, f.GetTime().AsTime(), activeTimeout)
+				default:
+					timer.Reset(activeTimeout)
+				}
+			} else {
+				// If there are no flows left, reset the timer to the default activeTimeout period.
+				timer.Reset(activeTimeout)
+			}
+
+		case flow := <-sm.flowChannel:
+			flowKey, err := sm.createFlowKey(flow)
+			if err != nil {
+				return err
+			}
+
+			// If the flow already exists in the cache, skip adding it again.
+			if _, found := sm.FlowCache.cache[flowKey]; found {
+				continue
+			}
+
+			// If the cache is full, evict the oldest flow to make room for the new flow.
+			// I think bufferSize should be renamed to maxBufferSize or something.
+			if sm.FlowCache.queue.Len() >= sm.FlowCache.bufferSize {
+				oldestElem := sm.FlowCache.queue.Front()
+				sm.FlowCache.queue.Remove(oldestElem)
+
+				// Process the evicted flow.
+				switch old := oldestElem.Value.(type) {
+				case *pb.FalcoFlow, *pb.CiliumFlow:
+					oldKey, err := sm.createFlowKey(old)
+					if err == nil {
+						delete(sm.FlowCache.cache, oldKey)
+						sm.sendNetworkFlowRequest(logger, old)
+					}
+				}
+			}
+
+			// Add the new flow to the cache and queue.
+			flowElem := sm.FlowCache.queue.PushBack(flow)
+			sm.FlowCache.cache[flowKey] = flowElem
+
+			// Reset the timer based on the new flow's timestamp.
+			switch f := flow.(type) {
+			case *pb.FalcoFlow:
+				resetTimer(timer, f.GetTimestamp().AsTime(), activeTimeout)
+			case *pb.CiliumFlow:
+				resetTimer(timer, f.GetTime().AsTime(), activeTimeout)
+			}
+		}
+	}
+}
+
+func resetTimer(timer *time.Timer, timestamp time.Time, timeout time.Duration) {
+	delay := timestamp.Add(timeout).Sub(time.Now())
+	// Not sure if I need this
+	// if delay <= 0 {
+	// 	delay = time.Second
+	// }
```

Comment:
No, you don't need this. Because if the new duration is negative it means that the flow has already expired, e.g. because we took too much time sending flows, etc.
That's ok, the timer will just fire immediately in this case, which is what we want. I quickly tested that.
```suggestion
```